### PR TITLE
Upgrade FieldPlane Averaging

### PR DESCRIPTION
**Pull request by gantech**
Template FieldPlaneAveraging and allow for average of horizontal velocity magnitude output

### CODE REVIEW

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 1]
```diff
@@ -89,16 +142,82 @@ private:
...
+    VelPlaneAveraging(CFDSim& sim, int axis_in):
+        FieldPlaneAveraging(sim.pde_manager().icns().fields().field
+                            , sim.time()
+                            , axis_in
+                            , true)
+    {
+        m_line_hvelmag_average.resize(m_ncell_line, 0.0);
+        if (m_comp_deriv)
+            m_line_hvelmag_deriv.resize(m_ncell_line, 0.0);
+    }
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:179]: Move this to `.cpp` file.

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 2]
```diff
@@ -89,16 +142,82 @@ private:
...
+
+    ~VelPlaneAveraging(){};
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:181]:
```suggestion
    ~VelPlaneAveraging() = default;
```

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 3]
```diff
@@ -89,16 +142,82 @@ private:
...
     template <typename IndexSelector>
     void
     compute_averages(const IndexSelector& idxOp, const amrex::MultiFab& mfab);
+    
+    /** fill derivatives of line storage */
+    void compute_line_derivatives();
+};
+
+template class FPlaneAveraging<Field>;
+template class FPlaneAveraging<ScratchField>;
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:161]: Move this to `.cpp` file

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 4]
```diff
@@ -25,13 +27,58 @@ namespace amr_wind {
...
-        amr_wind::Field& field, const amr_wind::SimTime& time, int axis_in = 2);
 
-    ~FieldPlaneAveraging(){};
+    /**
+    *  \param field_in [in] Field to be averaged
+    *  \param time [in] Time instance to determine output frequencies
+    *  \param axis_in [in] Direction along which planes are computed
+    *  \param compute_deriv [in] Should the derivative of the averages be computed
+    */
+    FPlaneAveraging(
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:41]: Move this to `.cpp` file

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 5]
```diff
@@ -25,13 +27,58 @@ namespace amr_wind {
...
+      if (m_comp_deriv)
+          m_line_deriv.resize(m_ncell_line * m_ncomp, 0.0);
+      m_line_xcentroid.resize(m_ncell_line);
+      
+      for (int i = 0; i < m_ncell_line; ++i) {
+          m_line_xcentroid[i] = m_xlo + (i + 0.5) * m_dx;
+      }
+    }
+    
+    ~FPlaneAveraging(){};
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:81]:
```suggestion
    ~FPlaneAveraging() = default;
```

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 6]
```diff
@@ -2,8 +2,10 @@
 #define FieldPlaneAveraging_H
 
 #include <AMReX_AmrCore.H>
+#include "amr-wind/CFDSim.H"
 #include "amr-wind/utilities/DirectionSelector.H"
 #include "amr-wind/core/Field.H"
+#include "amr-wind/core/FieldRepo.H"
 #include "amr-wind/core/SimTime.H"
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:9]: Could avoid including these headers

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 7]
```diff
@@ -2,8 +2,10 @@
 #define FieldPlaneAveraging_H
 
 #include <AMReX_AmrCore.H>
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:4]: header not needed?

File: amr-wind/utilities/FieldPlaneAveraging.H [hunk 8]
```diff
@@ -89,16 +142,82 @@ private:
...
+    /** fill derivatives of line storage */
+    void compute_line_derivatives();
+};
+
+template class FPlaneAveraging<Field>;
+template class FPlaneAveraging<ScratchField>;
+
+using FieldPlaneAveraging = FPlaneAveraging<Field>;
+
+//! Specific application of FieldPlaneAveraging to the velocity field
```
Comment (sayerhs) [inline amr-wind/utilities/FieldPlaneAveraging.H:165]: Use doxygen style comment and group directive as in `FieldPlaneAveraging`

**Comment by gantech** [comment]
Is there a reason why the DPC++ CI fails but the rest pass?

**Comment by sayerhs** [comment]
@gantech It appears to be an issue where some system configuration has changed and the github workflow file needs to be updated [see here](https://github.com/Exawind/amr-wind/runs/928507907?check_suite_focus=true). It is not a show-stopper for your pull request.

**Comment by gantech** [comment]
I have tested both ascii and netcdf outputs from the ABLStats class. I moved around the averaging process to the end of a time step. Could someone please verify the new order of doing things are ok?

**Comment by gantech** [comment]
I wanted to add @michaeljbrazell  to review the pull request. But it appears I cannot add reviewers to an existing pull request.

**Comment by sayerhs** [comment]
@gantech This is not failing in CUDA builds [see here](https://github.com/Exawind/amr-wind/runs/934343590?check_suite_focus=true#step:3:411)

File: amr-wind/wind_energy/ABLStats.cpp [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+              = gradT_arr(i,j,k,m_normal_dir);
+          tgrad_ptr[h2*m_ncells_h2 + h1].max_grad_loc = (k+0.5)*m_dn;
+        }
+      });
+    }
+
+    MPI_Reduce(tgrad_ptr, tgrad_g_ptr, m_ncells_h1 * m_ncells_h2,
+               MPI_2DOUBLE_PRECISION, MPI_MAXLOC,
+               amrex::ParallelDescriptor::IOProcessorNumber(),
+               amrex::ParallelDescriptor::Communicator() );
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.cpp:142]:
```suggestion
#ifdef AMREX_USE_MPI
    MPI_Reduce(tgrad_ptr, tgrad_g_ptr, m_ncells_h1 * m_ncells_h2,
               MPI_2DOUBLE_PRECISION, MPI_MAXLOC,
               amrex::ParallelDescriptor::IOProcessorNumber(),
               amrex::ParallelDescriptor::Communicator() );
#endif
```

We want to retain the option of not compiling with MPI enabled.

File: amr-wind/wind_energy/ABL.cpp [hunk 10]
```diff
@@ -83,59 +87,31 @@ void ABL::post_init_actions()
...
 void ABL::pre_advance_work()
 {
-    const auto& time = m_sim.time();
-
-    m_pa();
-
-    m_abl_wall_func.update_umean(m_pa);
+    
+    const VelPlaneAveraging& vel_pa = m_stats->vel_plane_averaging();
+    m_abl_wall_func.update_umean(vel_pa);
```
Comment (michaeljbrazell) [inline amr-wind/wind_energy/ABL.cpp:92]: I think the operator needs to be called here before update_umean

**Comment by gantech** [comment]
The plane averages now get computed at the start of the time step, but the integrated statistics like u*, w*, Q, zi are computed at the end of the time step. All regression tests now pass on Eagle compared to the latest dev golds on Eagle. If we're ok with this, could merge this pull request.

**Comment by sayerhs** [comment]
@gantech Please run `clang-format` before we merge this code.

**Pull request merged**